package.json

```diff
@@ -453,12 +453,14 @@
 		]
 	},
 	"scripts": {
+		"copy:assets": "cp node_modules/@highlightjs/cdn-assets/highlight.min.js media/highlightjs && cp -r node_modules/@highlightjs/cdn-assets/languages media/highlightjs && cp -r node_modules/@highlightjs/cdn-assets/styles media/highlightjs",
 		"vscode:prepublish": "npm run compile && npm run webpack",
 		"webpack": "npm run webpack-preview",
 		"webpack-preview": "webpack --mode production --config webpack.config.preview.js",
 		"webpack-preview-dev": "webpack --config webpack.config.preview.js --mode development",
 		"compile": "tsc -p ./",
 		"package": "vsce package",
+		"prepackage": "npm run copy:assets",
```

Comment:
Also if we include it in `compile` then it should function correctly in CI.

Reply:
My understanding is that `asciidoctor-vscode/.vscode/launch.json` is only used in local development.
We definitely want to run `npm run copy:assets` but I don't think we should run this task when we want to compile the code.

The task `prepackage` will run before the task `package` automatically. That's a built-in feature of npm scripts, see: https://docs.npmjs.com/cli/v7/using-npm/scripts#pre--post-scripts

Basically, when I run `npm run package` it will run the `prepackage` task first.

> Also if we include it in compile then it should function correctly in CI.

When we release a new version, we do:

https://github.com/asciidoctor/asciidoctor-vscode/blob/74bce1d292f36a1672d6df608d1aa4d1e8a85720/.github/workflows/release.yml#L40-L41

So, assets will be copied before creating the extension bundle.

> (Note: It seems like the workflow is not quite right at the moment because we currently have to manually call npm run webpack or npm run webpack-dev. Ideally a developer starting out wouldn't need to know this but we should at least not make it worse here I guess.)

Could you expand on that? We need to run `npm run webpack` after (or before?) `npm run compile`?

> ... or npm run webpack-dev

You meant `npm run webpack-preview-dev` ?
If the webpack task is only used in development I would probably always add the `--mode development` flag. Do you know if there are drawbacks? Maybe it's a bit slower to use the development mode?


To be honest, I think it would be better to consolidate the scripts and use a `npm run dev` or `npm start` to run/use the extension in local development.

Also, I don't think git commands/aliases belong to the `scripts`, so I would also remove `"prune-branches": "git remote prune origin"` and `"clean": "git clean -fdX"`.

Reply:
There were too many things to fix, I've created a new pull request: https://github.com/asciidoctor/asciidoctor-vscode/pull/452

Reply:
> The task prepackage will run before the task package automatically. That's a built-in feature of npm scripts, see: https://docs.npmjs.com/cli/v7/using-npm/scripts#pre--post-scripts

Thanks!

> There were too many things to fix, I've created a new pull request: #452

You are on :fire:, this is the most :heart: this extension has seen in many days. :+1: 

I locally merged this #452 on top of this PR to see if it all came together.

With these changes, the extension host doesn't run for me when I hit run/debug on the extension. I think this is because `npm watch` doesn't exit and the extension host expects it to (called via the prelaunch task in `launch.json`).

![image](https://user-images.githubusercontent.com/674783/134563812-6dc7fc86-3db4-4a81-99ce-aad3d8f39c97.png)

>Also, I don't think git commands/aliases belong to the scripts, so I would also remove "prune-branches": "git remote prune origin" and "clean": "git clean -fdX".

I think we should retain a clean task that removes build artefacts (in `media/highlightjs` and `dist`) but I agree that git commands are the wrong idea.

> You meant npm run webpack-preview-dev ?
> If the webpack task is only used in development I would probably always add the --mode development flag. Do you know if there are drawbacks? Maybe it's a bit slower to use the development mode?

The webpack commands compile the js which is run within the webview preview so the files generated in `dist` need to produced for either locally debugging/developing the extension and for packaging. If the `media/highlightjs` and `dist` files aren't present when the extension runs there will be errors in the webview because they don't exist.

When #452 is merged, I think we need to include `copy:assets` into the `dev` command and then I think it all works nicely. Both the `dist` folder and the `highlightjs` resources are then available for running locally:

```js
    "scripts": {
        "copy:assets": "cp node_modules/@highlightjs/cdn-assets/highlight.min.js media/highlightjs && cp -r node_modules/@highlightjs/cdn-assets/languages media/highlightjs && cp -r node_modules/@highlightjs/cdn-assets/styles media/highlightjs",
        "dev": "webpack --config webpack.config.preview.js --mode development && npm run copy:assets",
```

And then with the "prepackage" command everything should be created correctly for releases if I (now) understand correctly.